## scipy를 이용한 최적화

* scipy의 서브 패키지 optimize에 minimize 메소드를 이용하여 최적화한다
* 이 때 사용하는 BFGS 알고리즘이다
* minimize 메소드는 최적화를 시작하기 위한 초기값과 목적함수를 인수로 사용한다


#### * minimize 메소드
최적화 결과를 OptimizeResult 클래스 객체로 반환한다

    OptimizeResult 클래스의 속성들 :
    - x : 최적화 해
    - success : 최적화를 성공했을 때 True 반환
    - status : 종료 상태. 최적화에 성공했을 때 0을 반환
    - message : 최적화 성공/실패 메세지 문자열
    - fun : x 위치에서의 함수 값
    - jac : x 위치에서의 자코비안(그레디언트:기울기) 벡터 값
    - hess : x 위치에서의 헤시안 행렬 값
    - nfev : 목적 함수를 몇 번 호출한 횟수
    - njev : 자코비안 계산 횟수
    - nhev : 헤시안 계산 횟수
    - nit : x 이동 횟수

[참고] 데이타 사이언스 스쿨
https://datascienceschool.net/view-notebook/4642b9f187784444b8f3a8309c583007/


In [1]:
from scipy import optimize

# 목적함수
def f1(x):
    return (x-2)**2+2

# 초기값 설정
x0 = 0

result = optimize.minimize(f1, x0) # 목적함수와 초기값 지정
print(result)

      fun: 2.0
 hess_inv: array([[0.5]])
      jac: array([0.])
  message: 'Optimization terminated successfully.'
     nfev: 9
      nit: 2
     njev: 3
   status: 0
  success: True
        x: array([1.99999999])


[결과분석]

      fun: 2.0                              # 최적값의 해가 2인 경우 목적함수 값      
     hess_inv: array([[0.5]]) 
      jac: array([0.])
      message: 'Optimization terminated successfully.'  
     nfev: 9                                # 목적함수의 호출 횟수     
      nit: 2                                # x값의 위치를 2번 호출      
     njev: 3     
       status: 0                                # 종료   
      success: True                             # 성공종료  
        x: array([1.99999999])              # 최적값의 해(거의 2이다)
        
        
x값은 2번 이동했는데 목적함수를 9번 호출했다는 것은 자코비안값을 계산하기 위해(미분 계산?)
헤시안 값과 자코비안 벡터 값을 찾기 위해서 함수를 호출횟수가 늘어나는데
헤시안 값과 자코비안 값을 적정하게 선정하는 것이 좋다

## scipy의 서브 패키지(하위 모듈)


1. scipy.optimize : 최적화
2. scipy.stats : 통계
3. scipy.interplate : 보간법
4. scipy.io : 데이터 입출력
5. scipy.linalg : 선형대수 루틴 (linear algibura)